Accept header parsing and upgrade bug

r1mar opened the pull request:
Hello Zack,

ich have fixed one upgrade bug. The mongoose option reconnect tries doesn't exists any more. i deleted this from code.

I also improved the recognition of the Accept header. Multiple entries are now also supported and * placeholders.

xml and json are supported for the metadata. Only json is supported for other resources. If a mimetype is requested that is not supported, then 406 is returned.

Regards,
Richard

r1mar committed:
upgrade bug fixed

r1mar committed:
parsing of accept header optimized. closes #100

r1mar committed:
Merge branch 'zackyang000:master' into master

zackyang000 commented:
Hi r1mar,

1) Thanks for fix option bug.

2) Maybe we should have a default mimetype when the request doesn't contain a mimetype? Otherwise, you will not be able to use the browser to directly access resources. this is more commonly used in testing.

r1mar committed:
Tests for service document ready

r1mar committed:
Tests for default mimetypes. closes #99

r1mar committed:
new tests for mimetype defaults

r1mar committed:
name of new tests fixed

r1mar commented:
Hello zackyang000,

The protocol specifies xml as the default for the metadata. In principle, it is optional for data requests. However, since only json is implemented for, the choice is not large.

A test already existed for the metadata default. I moved this to a separate file and created another test for the default of the data requests.

In a nutshell. For the metadata the default is xml and for all other requests it is json.

If you don't fill the accept header, a response will always be sent. So it's not clear to me what you means.

Regards,
Richard

r1mar committed:
service document request implemented. closes #101

r1mar committed:
service document ready

r1mar commented:
Hello zackyang000,

I have implemented the service document request. I also discovered a bug in the data requests when specifying the mimetype "*/*". I fixed the bug immediately. Maybe that's what you meant.

zackyang000 commented:
Hi r1mar,

Sorry, maybe I didn't make it clear. I mean the defaults are for resource requests, the defaults are request headers.

E.g:
Use this code to start the service.
````
var odata = require('node-odata');
var server = odata('mongodb://localhost/my-app');
server.resource('Book', {
    title: string,
    Price: Quantity
});
server.listen(3000);
````

forward:
RUN `curl http://localhost:3000/books` will get the resource content.

back:
If RUN `curl http://localhost:3000/books -H 'Accept: application/json'` gives the same result, which is fine.
**But** RUN `curl http://localhost:3000/books` will get a 406 result.

So, I mean, should we create a default header (eg Accept: application/json) when the request doesn't include it?


Finally, I found it seems to be a bug? see [here](https://github.com/zackyang000/node-odata/commit/92a32e692032a57cbc34084afa4a69f814542b10#r86764668).

r1mar commented:
Hello zackyang000,

getMediaType is only called when any accept header is given. If a user requests a format that is not supported, they must also get a 406. The default branch doesn't work because of this. Curl sends automatically the header accept: */*. Support for asterix-header I implemented later here https://github.com/zackyang000/node-odata/pull/105/commits/2f2701319ae5da53a8e8a3bd81add48750a2d3d5 and one fix more here https://github.com/zackyang000/node-odata/pull/105/commits/c44872fdc1641927721aca68fa200cf09c34e072

You use for your test very old version of code. Can you test with last version of code? It should work.

r1mar committed:
Support for 'default Value', Maxlength and Boolean implemented. Also support for multiple mongo attributes for a property implemented

r1mar committed:
replacing dots in Names. Closes #106

r1mar committed:
batch operations implemented

r1mar committed:
batch works

r1mar committed:
Post give an error for missing _id and batch test not working

r1mar committed:
moved mocked tests to own folder

r1mar committed:
no _id error fixed

r1mar committed:
resolve hanging tests

r1mar committed:
generating exmaple for stackoverflow

r1mar committed:
partially support for enum arrays implemented

r1mar committed:
unbound Action works by direct call with metadata, but not per batch

r1mar committed:
implement support for actions in batch and workaround for ui5 bug with names of actions

r1mar committed:
fixed response bug in batch pipe

r1mar committed:
empty body bug in multipart/mixed

r1mar committed:
adding async support for authorization pipep. closes #84

r1mar committed:
transform hooks to middleware

r1mar committed:
validation of action parameters

r1mar committed:
Actions dokumented

r1mar committed:
Actions documented

r1mar committed:
Actions dokumentet

r1mar committed:
debug tracing implemented

r1mar committed:
remove unneccesary next call on the last hook

r1mar committed:
entity implemented

bitech-ag-rm committed:
Fix bug in multipart parser for get entity request. Extend registration of a mongo resource with options

bitech-ag-rm committed:
multipart-mixed 204 response in batch, DateTimeOffset to string implicit

bitech-ag-rm committed:
Splitting of odata and database logik. Validation of requests against the metadata

bitech-ag-rm committed:
[b] Splitting db and odata

bitech-ag-rm committed:
Splitting of db and odata logik completed

bitech-ag-rm committed:
cleaning and some bug fixes

bitech-ag-rm committed:
[b] 501 entity and 0 writer bug fixed

bitech-ag-rm committed:
first steps of singleton

bitech-ag-rm committed:
[b] mongoose singleton implemented

bitech-ag-rm committed:
[b] upgrade to node 18.17.0 and all dependencies

bitech-ag-rm committed:
upgarde to node 18.17.0 completed

bitech-ag-rm committed:
singleton implemented, hidden field attempt now to metadata, fix bug by xml presentation of metadata for array property

bitech-ag-rm committed:
singelton entity and annotations

bitech-ag-rm committed:
[b] Action parameters can be annotated

bitech-ag-rm committed:
[b] validation activated by mongo post and put

bitech-ag-rm committed:
extended odata error structure with target, detail. New Mongooses erro handler and handling mongoose validation issues. Sap annotation for error in details included

bitech-ag-rm committed:
fields annotation completed for action

bitech-ag-rm committed:
[b] writting null by selected nullable properties

bitech-ag-rm committed:
fixed bug of projektion in xml batches

bitech-ag-rm committed:
[b] patch for singleton, fix for deep projection, messages in success payload

bitech-ag-rm committed:
some fixes with deep selects and multipart mixed parser

bitech-ag-rm committed:
renaming project to nota

bitech-ag-rm committed:
0.1.0-alpha.0

bitech-ag-rm committed:
switch version and module name back

bitech-ag-rm committed:
Readme and example updated. multipart tests deleted

bitech-ag-rm committed:
start script path updated

r1mar commented:
Hello zackyang000,

I continued to extend of the OData implementation. The implementation is no longer dependent on MongoDB and can easily be used without a database or with another database. The metadata is no longer just used to publish the interface, but is also used for validation. There are completely new features such as batch requests, singletons and annotations. But we have a million breaking changes and the masochistic eslint validations are gone. ;-)

I am testing the implementation with the OpenUI5 framework client (https://openui5.hana.ondemand.com/). So far it's looking pretty good.

There is also a Yeoman Generator (https://www.npmjs.com/package/generator-men5) to generate apps with the men5 stack. For this I had to deploy a package myself. If you decide to accept this pull request, I will reference node-odata there.

Regards,
Richard

r1mar committed:
mapping refactored and client support for get implemented

r1mar committed:
client support for delete and count

zackyang000 commented:
Hi @r1mar,

It would be great to receive your PR submission, but I tried running the unit test locally and it failed. Am I missing some configuration?

<img width="1248" alt="image" src="https://github.com/zackyang000/node-odata/assets/3007674/9980b4b6-9a87-45a5-92bc-e40a016e22f6">

r1mar committed:
client support for singleton

r1mar committed:
fix first clone bug

r1mar commented:
Hello zackyang000,

I got the error now too. I had to delete the lib folder to do this. I deleted the line. You don't need the plugin because a mapping between the OData and the internal layer can be specified. This means that database fields can have different names than the OData properties of the entities.

Regards,
Richard

r1mar committed:
reorg of client tests, working of client support for list

r1mar committed:
client support for list request

r1mar committed:
client support for patch entity

r1mar committed:
client support implemented for patch, post and put

r1mar committed:
Client support documented
